### PR TITLE
async cacher

### PR DESCRIPTION
**Pull request by mapsam**
First pass at updating locking to manage and handle promise functions. This exports two methods:

1. `Locking` the original callback lock
2. `LockingAsync` a new async locker that can do things like this

```js
const { LockingAsync } = require('@mapbox/locking');
const got = require('got');

const jsonApi = new LockingAsync((url) => {
  return got(url).json();
});

const body = await jsonApi.get('https://api.mapbox.com');

console.log(jsonApi.stats); // { total: 3, hit: 2, miss: 1, locks: 2, refreshHit: 0, currentLocks: 0, size: 1 }
```

README updated and best viewed in https://github.com/mapbox/locking/tree/async

### CODE REVIEW

**Review by davidweimapbox — APPROVED** [review]
LGTM. great work.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,20 @@
 # Changelog
 
+## v4.0.0
+
+**BREAKING**
+
+- library now exports two items, `Locking` and `LockingAsync`
+- `Locking` function, which contains the same interface as previous versions
+- `LockingAsync` class, for managing async/promise functions of arbitrary parameters
+- update lur-cache 7.x, this includes updated options:
```
Comment (OzQu) [inline CHANGELOG.md:10]: `lur-cache` --> `lru-cache` ?

File: lib/async.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  }
+  */
+  #locks;
+  #stale;
+  #stats;
+
+  constructor(func, params) {
+    if (!func) throw new Error('Locking: locking class requires two arguments: function, params');
+    this.func = func;
+    this.cache = new LRU({ ...{ max: 1000 }, ...params });
```
Comment (OzQu) [inline lib/async.js:26]: Is destructuring `...{ max: 100 }` for some specific reason, wouldn't `new LRU({ max: 1000, ...params });` result with the same options object for LRU ?

**Review by OzQu — APPROVED** [review]
Looking good! I found one typo to fix from Changelog and one minor syntax question, but nothing blocking approval.

I didn't even know about this repositories existence, so it was great to read the PR through for learning!

File: lib/async.js (hunk 2 above)
Comment (mapsam) [reply lib/async.js]: Yep, looks like I accidentally doubled up the objects here. I'll simplify, thanks! Resolved in 2da4640

File: CHANGELOG.md (hunk 1 above)
Comment (mapsam) [reply CHANGELOG.md]: Resolved in 2da4640

**Comment by mapsam** [comment]
Thank you @OzQu!

**Pull request merged**